### PR TITLE
Label pane

### PR DESCRIPTION
**Pull request by nsrahmad**
See #605 for context.

An example to test:
```lisp
(eval-when (:compile-toplevel :load-toplevel :execute)
  (unless (find-package "CLIM")
    (ql:quickload 'mcclim)))

(in-package :clim-user)

(define-application-frame label-test ()
  ()
  (:panes
   (label :label-pane
	  :label "A Label")
   (int :interactor :width 600))
  (:layouts
   (default (vertically () label int))))

(define-label-test-command (com-change-label :name t :menu t)
    ((new-label string))
  (setf (clime:label-pane-label (find-pane-named *application-frame* 'label))
	new-label))

(defun run-label-test ()
  (run-frame-top-level (make-application-frame 'label-test)))
```
A sample run looks like:

anim

### CODE REVIEW

File: Examples/clim-fig.lisp [hunk 1]
```diff
@@ -427,7 +427,8 @@
       (T
        (erase-output-record latest-undo-entry *standard-output*)
        (push latest-undo-entry (clim-fig-redo-list *application-frame*))
-       (activate-gadget (find-pane-named *application-frame* 'clear))))
+       (activate-gadget (find-pane-named *application-frame* 'clear))
+       (activate-gadget (find-pane-named *application-frame* 'redo))))
```
Comment (scymtym) [inline Examples/clim-fig.lisp]: I have changes that synchronize the active/inactive state of these buttons with the corresponding commands (including menus). Should I submit those as a pull request?

Comment (nsrahmad) [reply Examples/clim-fig.lisp]: Sure, go ahead.

**Comment by scymtym** [comment]
Using a label instead of a text field in clim-fig is also a significant performance improvement.

File: Core/clim-core/panes.lisp [hunk 2]
```diff
@@ -2314,6 +2314,12 @@ SCROLLER-PANE appear on the ergonomic left hand side, or leave set to
    :text-style (make-text-style :sans-serif nil nil))
   (:documentation ""))
 
+(defmethod (setf clime:label-pane-label) :after (new-value (pane label-pane))
+  (let ((region (sheet-region pane)))
+    (multiple-value-bind (x1 y1 x2 y2) (rectangle-edges* region)
+      (draw-rectangle* pane x1 y1 x2 y2 :ink (pane-background pane)))
+    (repaint-sheet pane region)))
```
Comment (scymtym) [inline Core/clim-core/panes.lisp]: I would prefer just `(repaint-sheet pane (sheet-region pane))` here, but adding `(draw-rectangle* pane x1 (- ty a) x2 (+ ty d) :ink (pane-background pane))` before the `draw-text*` call in `handle-repaint`. That way, the label will always be drawn onto a freshly cleared background, no matter what caused the repaint. E.g. repainting the pane without changing the label would no longer lead to the label text appearing almost bold caused by overdrawing the anti-aliased glyph. This effect is visible in the "A" part of the label in the following screenshot:

screenshot

**Comment by dkochmanski** [comment]
Note that we have "always-repaint-background" mixin

On October 12, 2019 5:27:53 PM GMT+02:00, Jan Moringen <notifications@github.com> wrote:
>scymtym requested changes on this pull request.
>
>I would like to generalize the repainting improvement for `label-pane`.
>
>> @@ -2314,6 +2314,12 @@ SCROLLER-PANE appear on the ergonomic left
>hand side, or leave set to
>    :text-style (make-text-style :sans-serif nil nil))
>   (:documentation ""))
> 
>+(defmethod (setf clime:label-pane-label) :after (new-value (pane
>label-pane))
>+  (let ((region (sheet-region pane)))
>+    (multiple-value-bind (x1 y1 x2 y2) (rectangle-edges* region)
>+      (draw-rectangle* pane x1 y1 x2 y2 :ink (pane-background pane)))
>+    (repaint-sheet pane region)))
>
>I would prefer just `(repaint-sheet pane (sheet-region pane))` here,
>but adding `(draw-rectangle* pane x1 (- ty a) x2 (+ ty d) :ink
>(pane-background pane))` before the `draw-text*` call in
>`handle-repaint`. That way, the label will always be drawn onto a
>freshly cleared background, no matter what caused the repaint. E.g.
>repainting the pane without changing the label would no longer lead to
>the label text appearing almost bold caused by overdrawing the
>anti-aliased glyph. This effect is visible in the "A" part of the label
>in the following screenshot:
>
>screenshot

-- Wysłane za pomocą K-9 Mail.

**Comment by scymtym** [comment]
> Note that we have "always-repaint-background" mixin

That repaints the whole sheet region (and is currently only used in one place for some reason). The suggested change only repaints the region that could potentially be covered by the label.

File: Core/clim-core/panes.lisp (hunk 2 above)
Comment (nsrahmad) [reply Core/clim-core/panes.lisp]: Done. This is indeed better solution.

**Comment by scymtym** [comment]
Thank you.

**Comment by scymtym** [comment]
> Sure, go ahead.

#869

**Pull request merged**